distro/src/avail/Avail.avail/Foundation.avail/Late Math.avail

```diff
@@ -846,3 +868,247 @@ Public stable method "_(decimal)" is
  * @category "Mathematics" "Constants" "Numbers"
  */
 Macro "π" is [`3.141592653589793` ("π") → literal phrase];
+
+/**
+ * Compute the fused modular exponentiation of the base to the exponent, modulo
+ * the provided modulus. This method executes much faster than its unfused
+ * variant by leveraging iterated squaring and moduli.
+ *
+ * @method "_^_mod_"
+ * @param "base" "whole number"
+ * @param "exponent" "whole number"
+ * @param "modulus" "natural number"
+ * @returns "whole number"
+ *
+ * @category "Mathematics" "Integers" "Types"
+ */
+Public stable method "_^_mod_" is
+[
+	base : whole number,
+	exponent : whole number,
+	modulus : natural number
+|
+	moddedBase : whole number := base mod modulus;
+	exponentRemaining : integer := exponent;
+	result : whole number := 1;
+	Until exponentRemaining = 0 do
+	[
+		If exponentRemaining bit∧ 1 = 1 then
+		[
+			result := (result × moddedBase) mod modulus;
+		];
+		exponentRemaining := (exponentRemaining >> 1);
+		moddedBase := (moddedBase × moddedBase) mod modulus;
+	];
+	result
+];
+
+/**
+ * Restrict modular exponentiation to the combined restrictions upon the unfused
+ * exponentiation and modulus methods.
+ *
+ * @method "_^_mod_"
+ * @restricts "whole number's type"
+ * @restricts "whole number's type"
+ * @restricts "natural number's type"
+ */
+Semantic restriction "_^_mod_" is
+[
+	base : whole number's type,
+	exponent : whole number's type,
+	modulus : natural number's type
+|
+	(base ^ exponent) mod modulus
```

Comment:
Good catch, these restrictions are generally going to err to the "unreasonable" end of the spectrum.

> If the exponent bound is always < 2, intersect the base type with `[0..⌈modulus⌉)`

I don't think intersect is exactly the right operation, since disjoint base and modulus types should still have nonempty result types. (e.g. base `[10..12]` mod `[4..5]`'s has possible values in `[0..3]`).

Stepping back, this method is basically designed for trying to make larger bases and exponents calculable, so I'm not sure that returning anything stricter than `[0..⌈modulus⌉)` is actually going to have real world benefit..? I'll try to sketch the cases out though.

Reply:
True dat!  `"_mod_"` has a pretty good semantic restriction for the case that the exact divisor is known.  It checks if the quotients of the largest dividend (over divisor) and the smallest dividend (over divisor) are equal, and if so, it computes the two remainders and uses the range spanning them.  Otherwise it uses [0..divisor), since both 0 and divisor-1 must both be possible results in that case.

I suppose you could generalize that to a variable base, but I'm sure you're right about diminishing benefits beyond `[0..⌈modulus⌉)`.  Besides, multiple semantic restrictions can make varying claims about the resulting type, and their intersection is used for the expected return type at a call site, so someone could always add a stricter case if they felt like coding it themselves.